Content/SCPCB/Game/Rooms/Room.as

```diff
@@ -0,0 +1,189 @@
+// # Room::Type ----
+namespace Room { enum Type {
+    R1 = 1,
+    R2 = 2,
+    R2C = 3,
+    R3 = 4,
+    R4 = 5
+} }
+
+// # Room::Model@ ----
+namespace Room { class Model {
+	string path;
+	Model() {}
+	Model(string&in iPath) { path=iPath; }
+	Game::Room@ instantiate() { return Game::Room(path); }
+} }
+
+// # Room::Model@ ----
+namespace Room { class ModelCBR : Model {
+	ModelCBR(string&in iPath) {super(); path=iPath; }
+	Game::Room@ instantiate() { return Game::RoomCBR(path); }
+} }
+
+
+// # Room::Sound@ ----
+namespace Room { class Sound {
+	string path;
+	Sound() {}
+	Sound(int i) {} // picksoundid
+	Sound(string&in iPath, float&in iScale, string&in iSkin="") { path=iPath; }
+} }
+
+// # Room::Icon@ and Room::Icon::Model@ ----
+namespace Room {
+	// Room::Icon@ ----
+	class Icon : Util::Icon { Icon() {super();} Icon(string&in texStr) {super(texStr);} Icon(Texture@&in tex) {super(@tex);} }
+
+	// Room::Icon::Model@ ----
+	namespace Icon { class Model : Util::Icon::Model {
+		Model(string&in iPath, float&in iScale, Vector3f&in iRotation, Vector2f&in iPos, string&in iSkin="") {super(iPath,iScale,iRotation,iPos,iSkin);}
+		Model(string&in iPath, Vector3f&in iScale, Vector3f&in iRotation, Vector2f&in iPos, string&in iSkin="") {super(iPath,iScale,iRotation,iPos,iSkin);}
+	} }
+}
+
+
+// # ??? : Room::Template@ ----
+namespace Room { interface TemplateInterface { Room@ instantiate(); } }
+namespace Room { abstract class Template : Room::TemplateInterface {
+	Room@ instantiate() {return null;}
+	Template() { Room::templates.insertLast(@this); }
+	void internalConstruct() {
+		Debug::log("Internal constructed " + name);
+		@mesh=model.instantiate();
+		construct();
+	}
+	void construct() {} // override
+
+	string name = "SCP-000";
+	string zone = "000";
```

Comment:
Using strings as identifiers at runtime is more often than not a mistake, and certainly a mistake here. We need to discuss the rationale behind this over the use of the Zone class.

Reply:
This is more or less a placeholder that does nothing, except output a debug string while loading.

I am interested to know more about what you mean by runtime identifiers.
I have used them in several places as default initializers, but I can move them to their respective constructor functions.
I admit, I have noticed some occasional strange behavior when variables are given predefined values like that.

The zone class, and rooms system in general, is another matter that has a rather reasonable rationale.
That being, the zone doesn't actually matter. the room you're standing in, is what actually determines the zone you're in.
It does not need to be a straight line along the map, it could be a deformed shape.
There's some other factors about rendering, and how adjacency, and linear adjacency, is a lot more important than anything else related to rooms, including zones. Or at least I think so. Happy to discuss this further!

Reply:
> That being, the zone doesn't actually matter. the room you're standing in, is what actually determines the zone you're in.

You seem to misunderstand the purpose of the Zone class. The Zone class is meant to represent a lot more than simply a collection of rooms it contains. It also represents the way that those rooms are placed, what shared state is applicable to only a certain zone, and effectively the entire map when the zone is loaded.

Reply:
I suspected zone higher functions would be desired.
I did not want to iterate further on anything related to the rooms, room spawning, world generation, or anything like that, because the registration system is not yet approved or finalized, so building off it would be counter-productive.
This is partly why I've slowed down and wrote up a description of what i've done so far, and want to focus on what's already done so we can have these discussions before I take anything too much further.

It is currently just a placeholder string until I understand what your design intentions and goals there are.
The room objects are currently intended only to spawn instances of the .cbr meshes into the gameworld, so it is not developed far beyond that.